editor/utils/dom.js

```diff
@@ -56,18 +57,85 @@ export function isEdge( container, start = false ) {
 }
 
 /**
- * Places the caret at start or end of a given element
+ * Holds the first caret boundary rectangle for subsequent calls.
+ *
+ * @type {DOMRect}
+ */
+let firstVerticalRect;
+
+/**
+ * Resets the first caret position.
+ */
+export function resetVerticalPosition() {
+	firstVerticalRect = null;
+}
+
+/**
+ * Check whether the caret is vertically at the edge of the container.
  *
- * @param  {Element} container DOM Element
- * @param  {Boolean} start     Position: Start or end of the element
+ * @param  {Element} container Focusable element.
+ * @param  {Boolean} reverse   Set to true to check top, false for bottom.
+ * @return {Boolean}           True if at the edge, false if not.
  */
-export function placeCaretAtEdge( container, start = false ) {
+export function isVerticalEdge( container, reverse ) {
+	if ( [ 'INPUT', 'TEXTAREA' ].indexOf( container.tagName ) !== -1 ) {
+		return isHorizontalEdge( container, reverse );
+	}
+
+	if ( ! container.isContentEditable ) {
+		return true;
+	}
+
+	const selection = window.getSelection();
+	const range = selection.rangeCount ? selection.getRangeAt( 0 ) : null;
+
+	if ( ! range || ! range.collapsed ) {
+		return false;
+	}
+
+	// Adjust for empty containers.
+	const rangeRect =
+		range.startContainer.nodeType === window.Node.ELEMENT_NODE ?
+		range.startContainer.getBoundingClientRect() :
+		range.getClientRects()[ 0 ];
```

Comment:
Minor point, but I find that multi-line ternaries read better like:

```js
a
  ? b
  : c
```

but currently our linter has no preference.

Reply:
Fair enough. I've been used to this way. Maybe we should add a rule so that it's the same everywhere?

Reply:
My only worry is that the Javascript interpreter of some browser out there will see the condition as the end of the statement if the question mark is not the same line. It may be changed in transpiling though so if Babble gets it right then I suppose I don't need to worry.

Reply:
That's probably why I've been used to it this way.

Reply:
I had a whole reply with an excerpt of the [ECMA spec](https://www.ecma-international.org/ecma-262/8.0/index.html#sec-automatic-semicolon-insertion) and an analysis making the case that only a faulty browser would be too eager to insert a semicolon in this case, but then I found the right Crockford [source](http://javascript.crockford.com/code.html). :)

> The ternary operator can be visually confusing, so ? question mark always begins a line and increase the indentation by 4 spaces, and : colon always begins a line, aligned with the ? question mark. The condition should be wrapped in parens.

My take is that Crockford has historically been a conservative reference on JS (and, [apparently](https://stackoverflow.com/a/7259181), a previous version of the same text had advised against having an operator at the beginning of a line), and thus his OK means that we're totally safe here and that it's ultimately a stylistic decision. How do y'all weigh in?

Reply:
I don't have a strong opinion. If browsers are a worry, we anyway process with Babel and everything gets minified.